### PR TITLE
Optimize Factorize Linear Constraints

### LINKED ISSUE #1011: Optimize Factorize Linear Constraints
Currently, in `factorize_linear_constraints` function, we take out the fixed parameters but this applies only to `np.where(np.count_nonzero(A, axis=1) == 1)`. This is good but since we use the null-space a lot later on the code, I think we should make it as accurate as possible. What I have in mind is something like this,
```python 
    indices_idx = np.arange(1, objective.dim_x + 1)
    indices_row = np.arange(1, A.shape[0] + 1)
    while len(np.where(np.count_nonzero(A, axis=1) == 1)[0]):
        # fixed just means there is a single element in A, so A_ij*x_j = b_i
        fixed_rows = np.where(np.count_nonzero(A, axis=1) == 1)[0]
        print(f"Number of Fixed rows: {len(fixed_rows)}")
        # indices of x that are fixed = cols of A where rows have 1 nonzero val.
        _, fixed_idx = np.where(A[fixed_rows])
        unfixed_rows = np.setdiff1d(np.arange(A.shape[0]), fixed_rows)
        unfixed_idx = np.setdiff1d(np.arange(objective.dim_x), fixed_idx)

        # find the global index of the unfixed variables
        indices_idx[fixed_idx] = 0  # set fixed indices to 0
        indices_idx = indices_idx[
            np.nonzero(indices_idx)
        ]  # remove 0s, hence fixed indices
        indices_row[fixed_rows] = 0  # set fixed indices to 0
        indices_row = indices_row[
            np.nonzero(indices_row)
        ]  # remove 0s, hence fixed indices

        if len(fixed_rows):
            # something like 0.5 x1 = 2 is the same as x1 = 4
            b = put(b, fixed_rows, b[fixed_rows] / np.sum(A[fixed_rows], axis=1))
            A = put(
                A,
                Index[fixed_rows, :],
                A[fixed_rows] / np.sum(A[fixed_rows], axis=1)[:, None],
            )
            xp = put(xp, fixed_idx, b[fixed_rows])
            # Some values might be fixed, but they still show up in other constraints
            # this is where the fixed cols have >1 nonzero val.
            # For fixed variables, we delete that row and col of A, but that means
            # we need to subtract the fixed value from b so that the equation is balanced.
            # e.g., 2 x1 + 3 x2 + 1 x3 = 4 ; 4 x1 = 2
            # combining gives 3 x2 + 1 x3 = 3, with x1 now removed
            b = put(
                b,
                unfixed_rows,
                b[unfixed_rows] - A[unfixed_rows][:, fixed_idx] @ b[fixed_rows],
            )
        A = A[unfixed_rows][:, unfixed_idx]
        b = b[unfixed_rows]
```
So, for example,
```math
\begin{align}
3a+4b+c &= 5 \\
a - b &= 1 \\
a &= 2  \rightarrow  b=1 \rightarrow c = -5
\end{align}
```
But current code will stop after setting a=2 and b=1.

This should clean-up the whole fixed parameters from calculations. I think this should help a lot to Poincare branch since additional optimizable parameters cause A matrix to be much bigger and `Ainv_full, Z = svd_inv_null(A)` is less and less accurate. I don't suggest whole Gaussian elimination, but I think fixed values should never be included in the calculations. 

By the way, above code has a bug. The last array update procedure doesn't work properly. Instead of updating the A and b during while loop, it only updates it at the end so, first loops use original A matrix for some reason.

### PR DESCRIPTION
**Pull request by YigitElma**
Resolves #1011 

I just made the fix parameter removal a loop such that we remove every explicitly and implicitly fixed parameter before SVD. The self-consistency part can be done on a separate PR, I will test it myself. Some explanation why I did this is [here](https://www.overleaf.com/read/mjrrsbyztyhm#9a79da)

### CODE REVIEW

File: desc/objectives/utils.py [hunk 1]
```diff
@@ -97,34 +97,61 @@ def factorize_linear_constraints(objective, constraint):  # noqa: C901
...
-        # e.g., 2 x1 + 3 x2 + 1 x3 = 4 ; 4 x1 = 2
-        # combining gives 3 x2 + 1 x3 = 3, with x1 now removed
-        b = put(
-            b,
-            unfixed_rows,
-            b[unfixed_rows] - A[unfixed_rows][:, fixed_idx] @ b[fixed_rows],
-        )
-    A = A[unfixed_rows][:, unfixed_idx]
-    b = b[unfixed_rows]
+    # will store the global index of the unfixed rows with shifted indices by 1
```
Comment (dpanici) [inline desc/objectives/utils.py:100]: why shift the indices by one?

Comment (YigitElma) [reply desc/objectives/utils.py:100]: That is actually related to how I delete fixed indices later in the code. For some reason, I chose to set the fixed parameters to 0 and then delete the 0 values, so the initial array shouldn't have 0. At the time I wrote it, it felt convenient to use np.nonzero() to find nonzero elements. If it is hard to follow, I can change it to something else

Comment (YigitElma) [reply desc/objectives/utils.py]: I used np.delete now

**Review by dpanici — APPROVED** [review]

**Comment by dpanici** [comment]
we'll wait for the benchmark to run again though as I wonder if it does make a difference. And for a 2nd reviewer as usual

File: desc/objectives/utils.py [hunk 2]
```diff
@@ -97,34 +97,52 @@ def factorize_linear_constraints(objective, constraint):  # noqa: C901
...
+            # we need to subtract the fixed value from b so that the equation is
+            # balanced.
+            # e.g., 2 x1 + 3 x2 + 1 x3 = 4 ; 4 x1 = 2
+            # combining gives 3 x2 + 1 x3 = 3, with x1 now removed
+            b = put(
+                b,
+                unfixed_rows,
+                b[unfixed_rows] - A[unfixed_rows][:, fixed_idx] @ b[fixed_rows],
+            )
+        A = A[unfixed_rows][:, unfixed_idx]
```
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/utils.py:143]: ie, suppose row 3 is fixed, and row 5 will be fixed once row 3 is removed. On the first iteration, row 3 is removed, but that means what was row 5 is now row 4?

[the remark this answers is not in the document]
Comment (YigitElma) [reply desc/objectives/utils.py:143]: Yes, I handle that change with the `global_fixed_idx` for `xp` inside the loop, since that is the only array that doesn't change its size. And after the loop, I use `indices_idx` which holds the global indices of the whole unfixed parameters.

[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/objectives/utils.py:143]: I'm not sure that fixes the issue. `global_fixed_idx = indices_idx[fixed_idx]`, but `fixed_idx` depends on `A` which is changing size?

File: desc/objectives/utils.py [hunk 3]
```diff
@@ -97,34 +97,52 @@ def factorize_linear_constraints(objective, constraint):  # noqa: C901
...
-        b = put(
-            b,
-            unfixed_rows,
-            b[unfixed_rows] - A[unfixed_rows][:, fixed_idx] @ b[fixed_rows],
-        )
-    A = A[unfixed_rows][:, unfixed_idx]
-    b = b[unfixed_rows]
+    # will store the global index of the unfixed rows, idx
+    indices_row = np.arange(A.shape[0])
+    indices_idx = np.arange(A.shape[1])
```
[the remark this answers is not in the document]
Comment (YigitElma) [reply desc/objectives/utils.py:102]: Yeah, I tried to be consistent with variable names like `fixed_idx`
